### ISSUE TITLE
Loop support in imported wav files

### ISSUE THREAD

**Issue by CurlyMC** [surge-synthesizer/surge#2777]
It would be fantastic if the in the wavetable oscillator imported wave files, could be looped.
In waveosaur its possible to create loop points inside an audio wave file.
Would be very awesome if Surge could use these loop points to loop a part of the wavefile.
Thanks to Claes Johanson and the Surge Synth Team for this fantastic vst instrument !!

**Comment by baconpaul** [comment]
You mean in one-shot mode if they would loop? Or you mean the import would read them as tables more easily?

We do use loop points to try and glean tables.

I'm glad you are enjoying surge!

**Comment by CurlyMC** [comment]
Dear Paul, yes I mean in one-shot mode ( as imported audio file thats just played as is in Surge, with a loop ).

**Comment by baconpaul** [comment]
Gotcha.

I'm in the middle of another bit of the code now but let me tag this to look at later in the week. I think it is pretty easy - just add a flag to the internal data structure and don't run off the end if the right markers are in place - but I'm not sure it would make our 1.8.0 milestone (which has loads of new features! and so we are trying to avoid adding new ones.)

**Comment by CurlyMC** [comment]
I understand !! 
Thank you for the fast reply, Kind Regards.

**Comment by mkruselj** [comment]
TBH I would much rather add a sample oscillator instead of trying to push the WT oscillator to behave as a sample oscillator with those oneshot workarounds and whatnot...

**Comment by CurlyMC** [comment]
@mkruselj ; indeed, I couldn't agree more.
And a sample oscillator with added sfz support would be a fantastic future addition for Surge 👍 
Since I'm not a programmer, maybe basic audiofile loop functionality could be more easy to implement.

**Comment by baconpaul** [comment]
Even though you are not a programmer, you gleaned the ease of implementation difference correctly :)

**Comment by CurlyMC** [comment]
https://www.wavosaur.com/quick-help/loop-points-edition.php

**Comment by mkruselj** [comment]
You've alredy mentioned that. Bumping an issue won't make it happen any faster 😉 

Surge currently uses the loop points in WAV files to know the size of wavetable frames in samples (which is one of legitimate ways to produce a wavetable). Since this is a valid wavetable format (as in - people might actually have wavetables created like this and would want to use them as wavetables in Surge, rather than looped oneshots), it makes it incompatible with using loop points for any other purpose, as far as I can see. You would need to use Surge's .wt format which has an actual "oneshot" flag that marks the sample to be played back not as a wavetable but as a oneshot.

**Comment by CurlyMC** [comment]
Hi @mkruselj & Surge team, thank you for the reply and explanation, I appreciate !
I did post that wavosaur looppoints link just to give some additional information, because i did mention the wav loop support is 1.82 listed.
If it's giving the programming-team feeling of bumping the issue, perfect..... but it wasn't my intention :-)

As far as i know ( as an wavosaur user ) there can be much markers but only 1 loop inside an audiowave file.
These loopstart and loop-end points created in wavosaur are embedded inside the waveheader.
Interesting to hear that Surge is using looppoints to determine wavetable frame size. 

When I import a wav file, a sample from a few seconds - also not a wavetable - with these looppoints inside Highlife = retro vst instrument, it works perfectly.
Highlife recognizes the in wavosaur set looppoints and the wave file is played and looped in the vst instrument ( http://vst4you.com/pages/vst%20instruments/highlife.html ).
These wave files with a partial loop section, lets say a few seconds audio snippets ( for example nature-, voices- or other non synthetic soundsamples ) could add an extra powerful dimension to Surge's already immense sonic capabilities.
I want to emphasize that's not my intention to play a large 1 minute wave files, but just small snippets with a loop for non synthetic sounds. 

Your explanation of doing this with one shot wt files, makes me confused.
I dont have succes in looping one shot wt files, and I also think we should distinguish wave and wt files.
Problem is also that wavosaur can't create these oneshot or other wt-format files, but it's super easy to create wave files with looppoints in wav-file format.

As already suggested maybe there can be an additional 'sample oscillator', especially for dropping in wave files for this specific workflow, or a popup like @baconpaul did mention once.
If there's any basic possibility to play a partially looped the wav-file someway, it would be awesome !
Maybe in Surge 'sample oscillator' just 2 fields with loopstart- and loopend samplerange, and an option to play the loopsection forward, backwards ect. could do the trick.
Something like the settings in the loop properties window mentioned in the wavesaur manual link.

Hoping i did explain it as clear as possible, without do harm to anybody :-).
Want to emphasize that my comment is meant on a possitive constructive way, just like your feedback ok
Greetz, Luc.

**Comment by mkruselj** [comment]
Highlife is an actual sampler, not a wavetable synth though. Of course it would understand the loop points differently. While wavetables _are_ samples, they don't behave like an audio file behaves in a sampler. So loop points have a different purpose in Surge, as I explained, which caters more to wavetable usage.

The real proper solution for this is indeed a sampler/sfz oscillator. But I am not sure we really want this for Surge, when we have [another pretty cool project](https://github.com/surge-synthesizer/shortcircuit3) slated for this year that includes samples. 😉 

Also, regarding .wt files, you can use AudioTerm to generate them, or you can use Python scripts that we have [here](https://github.com/surge-synthesizer/surge/tree/main/scripts/wt-tool).


Thanks for your feedback, though.

**Comment by CurlyMC** [comment]
ShortCircuit 3 ? Wow thats big news, interesting.
Still hoping Surge basic wave file play and looping ( and maybe sfz ) could get some love from the programmers !
Otherwise I dont exactly understand why this 'Loop support in imported wav files' is on the 1.8.2 list, maybe someone can explain ( maybe I did misunderstand this one-shot mode ? ).
Thanks for the fast reply @mkruselj  !!

**Comment by baconpaul** [comment]
I am also not sure why this is in the 182 list!! I don’t think there’s something I can do here for that release. @mkruselj am I missing something?

**Comment by mkruselj** [comment]
I pushed it to 1.9.0 some time ago then forgot after we renamed 1.9.0. to 1.8.2 to push it out. Pushed it to Currently unscheduled 😃

**Comment by CurlyMC** [comment]
@baconpaul @mkruselj ; Unfortunately to hear that this feature has been postponed for now, but grateful for all future development for the Surge Synthesizer ! Keep it going on.
Kind Regards, Luc.

**Comment by baconpaul** [comment]
Yeah the book of work is huge and we sort of have to focus on some architectural things before we can do a proper job on things like importers and expanded oscillators

that said I’m less opposed to an sfz player in surge than @mkruselj is ! Especially now we have the sc3 code.... But we only have so many hands 

stay tuned though.

**Comment by mkruselj** [comment]
Closing this issue in favor of #3436.
